crates/pop-drink/src/error.rs

```diff
@@ -0,0 +1,297 @@
+//! A set of errors used for testing smart contracts.
+
+use std::fmt::Debug;
+
+pub use drink::{
+	pallet_assets::Error as AssetsError, pallet_balances::Error as BalancesError,
+	pallet_contracts::Error as ContractsError,
+};
+use scale::{Decode, Encode};
+
+fn decode<T: Decode>(data: &[u8]) -> T {
+	T::decode(&mut &data[..]).expect("Decoding failed")
+}
+
+/// Runtime error for efficiently testing both runtime module errors and API errors.
+/// It is designed for use with the `assert_err!` macro.
+///
+/// # Generic Parameters
+///
+/// - `ModuleError` - Error type of the runtime modules. [Reference](https://paritytech.github.io/polkadot-sdk/master/solochain_template_runtime/enum.Error.html).
+/// - `ApiError` - Error type of the API, which depends on version. [Reference](https://github.com/r0gue-io/pop-node/tree/main/pop-api).
+/// - `MODULE_INDEX` - Index of the variant `Error::Module`. This is based on the index of [`ApiError::Module`](https://github.com/r0gue-io/pop-node/blob/main/primitives/src/lib.rs#L38).
+///
+/// # Examples
+///
+/// ### Runtime module errors
+///
+/// - Import types to construct runtime module errors:
+///
+/// ```rs
+/// use drink::devnet::{
+/// 	Assets,
+/// 	AssetsError::AssetNotLive,
+/// 	Balances::BelowMinimum,
+/// 	BalancesError::BelowMinimum
+/// };
+/// ```
+///
+/// - Construct a runtime module error [`Assets(AssetNotLive)`](https://paritytech.github.io/polkadot-sdk/master/pallet_assets/pallet/enum.Error.html#variant.AssetNotLive):
+///
+/// ```rs
+/// Error::Module(Assets(AssetNotLive))
+/// ```
+///
+/// - Construct a runtime module error [`Balances(InsufficientBalance)`](https://docs.rs/pallet-balances/latest/pallet_balances/pallet/enum.Error.html#variant.InsufficientBalance):
+///
+/// ```rs
+/// Error::Module(Balances(InsufficientBalance))
+/// ```
+///
+/// ### API errors
```

Comment:
```rs
	/// Error type of the runtime modules. [Reference](https://paritytech.github.io/polkadot-sdk/master/solochain_template_runtime/enum.Error.html).
	Module(ModuleError),
	/// Every [`ApiError`](https://github.com/r0gue-io/pop-node/blob/52fb7f06a89955d462900e33d2b9c9170c4534a0/primitives/src/lib.rs#L30).
	Api(ApiError),
```
it is understandable that ModuleError also comes from Api Error, but from a high level user's perspective, the module errors are what defined here [Module Error](https://paritytech.github.io/polkadot-sdk/master/solochain_template_runtime/enum.Error.html) which follows the format .e.g., `Assets(NoAccount)`

`ApiError` is a layer lower than the `ModuleError` which is more like raw error and also covers more cases on the runtime level while `ModuleError` is only for modules. 

You have any recommendation on the naming stuffs? Because if we do not agree with the ApiError we also need a better way to express it in the Error type as well. 

```
Error::Raw?
```

Reply:
Resolved by removing all the introduction of `ApiError`
```
ApiError -> RuntimeCallError
```
I changed it to `RuntimeCallError` and don't mention anything about errors returned by the API. I reordered the `RuntimeCallError` and `ModuleError` so developers will read the `RuntimeCallError` description first before visiting `ModuleError`. From my perspective, `ModuleError` is simply a subset of reasons why the runtime call fails, originating from the modules.  

Hence, both errors types `RuntimeCallError` and `ModuleError` come from the runtime. Simply speaking, `RuntimeCallError` is a parent of `ModuleError`, but `ModuleError` helps to test the runtime module error in a better way.